x-pack/plugins/task_manager/server/ephemeral_task_lifecycle.ts

```diff
@@ -0,0 +1,200 @@
+  config: TaskManagerConfig;
+  middleware: Middleware;
+  elasticsearchAndSOAvailability$: Observable<boolean>;
+  pool: TaskPool;
+  lifecycleEvent: Observable<TaskLifecycleEvent>;
+}
+
+export type EphemeralTaskInstanceRequest = Omit<EphemeralTaskInstance, 'startedAt'>;
+
+/**
+ * The public interface into the task manager system.
+ */
+export class EphemeralTaskLifecycle {
+  private definitions: TaskTypeDictionary;
+  private pool: TaskPool;
+  private lifecycleEvent: Observable<TaskLifecycleEvent>;
+  // all task related events (task claimed, task marked as running, etc.) are emitted through events$
+  private events$ = new Subject<TaskLifecycleEvent>();
+  private ephemeralTaskQueue = new Set<EphemeralTaskInstanceRequest>();
+  private logger: Logger;
+  private config: TaskManagerConfig;
+  private middleware: Middleware;
+
+  /**
+   * Initializes the task manager, preventing any further addition of middleware,
+   * enabling the task manipulation methods, and beginning the background polling
+   * mechanism.
+   */
+  constructor({
+    logger,
+    middleware,
+    definitions,
+    pool,
+    lifecycleEvent,
+    config,
+  }: EphemeralTaskLifecycleOpts) {
+    this.logger = logger;
+    this.middleware = middleware;
+    this.definitions = definitions;
+    this.pool = pool;
+    this.lifecycleEvent = lifecycleEvent;
+    this.config = config;
+
+    this.lifecycleEvent
+      .pipe(
+        filter(
+          // we want to know when the queue has ephemeral task run requests
+          (e) => {
+            return (
+              this.ephemeralTaskQueue.size > 0 &&
+              e.type === TaskEventType.TASK_POLLING_CYCLE &&
+              // and when a polling cycle has just completed,
+              // (e.type === TaskEventType.TASK_POLLING_CYCLE ||
+              //   // or the "load" in the TaskPool has changed (meaning a task has just completed)
+              //   (e.type === TaskEventType.TASK_MANAGER_STAT && e.id === 'load')) &&
+              this.getCapacity() === 0
+            );
+          }
+        )
+      )
+      .subscribe(async (e) => {
+        for (const task of this.ephemeralTaskQueue.values()) {
+          const event = asEphemeralTaskDelayedDueToCapacityEvent(task.id, asOk(task));
+          this.emitEvent(event);
+          this.ephemeralTaskQueue.delete(task);
+        }
+      });
+
+    this.lifecycleEvent
+      .pipe(
+        filter(
+          // we want to know when the queue has ephemeral task run requests
+          (e) =>
+            this.ephemeralTaskQueue.size > 0 &&
+            e.type === TaskEventType.TASK_POLLING_CYCLE &&
+            // and when a polling cycle has just completed,
+            // (e.type === TaskEventType.TASK_POLLING_CYCLE ||
+            //   // or the "load" in the TaskPool has changed (meaning a task has just completed)
+            //   (e.type === TaskEventType.TASK_MANAGER_STAT && e.id === 'load')) &&
+            this.getCapacity() > 0
```

Comment:
I was listening to the concern @chrisronline had in the sync, which was basically - what happens if the ephemeral queue is filling up and Task Manager keeps pulling tasks our of ES.

I think this can be addressed pretty well by ensuring Task Manager prioritises the use of capacity for Ephemeral tasks, and the key to that (I think) is in this part of the code here.

What we're doing here is the following:
We listen for each polling cycle event and if there's capacity we pull in ephemeral tasks.
I did this in the spike *solely* to validate my theory, not because I thought that's how it should work.

What we should be doing here is pulling in ephemeral tasks every time:

1. A polling cycle completes (we already do that)
2. A task of any kind completed

I believe that should be enough to ensure that the moment a task completes, it will pull in an ephemeral task, and usually that would cause a chain of ephemeral tasks triggering each other one after the other (or as many as the capacity allows in parallel). Assuming they're running at the moment a next polling cycle is triggered - there should be no available workers for tasks from ES until the ephemerals are drained.